### DISCUSSION TITLE
Add details for Signature Verification -- Webhooks page

### DISCUSSION THREAD

**Discussion by parsa-asgari** [outline/outline#6448] [Ideas]
Hi,

I have an idea for a small improvement for the [Webhooks](https://docs.getoutline.com/s/guide/doc/webhooks-gB7HYhS6yq) documentation page.
Currently, the documentation does not explain the way one could verify the signature. In other words, it does not specify the pattern by which one should make the string with which the secret could be used to generate an HMAC.

Therefore, it would be great if you could add it.

As I could not find the documentation repo to contribute to, here is a basic Python implementation(based on [This Link](https://github.com/outline/outline/blob/56d05a2595dd6eae636078599d0239fd05e2d0db/server/models/WebhookSubscription.ts#L163)):
```python
######
# The pattern is something like "[timestamp].[payload]" assuming that:
#      [timestamp]: the timestamp specified in the 'outline-signature' header
#      [payload]: the body of the request in string format
######
import hmac
import hashlib


# example signature:  't=1706609916240,s=45b23d418b5d5c06af7721b526170d23d8699d3de4fc499a21dc33e3576b8236'
outline_signature_timestamped = request.headers['outline-signature']
timestamp = outline_signature_timestamped.split(',')[0].split('=')[1] 
signature = outline_signature_timestamped.split(',')[1].split('=')[1] 
body = request.body()

SECRET= 'YOUR_OUTLINE_SIGNATURE'
DATA = f'{timestamp}.{body.decode()}'
h = hmac.new( SECRET.encode(), DATA.encode(), digestmod=hashlib.sha256 )
if h.hexdigest() == signature:
  print('Signatures match.😝')
else:
  print('Signatures do not match.😖')

```

**Comment by tommoor** [comment]
Definitely need an example on there, given we're a end-to-end typescript project I think it would make sense to use TS/JS though.

**Reply by parsa-asgari** [reply]
Sorry for the delay. Here is the code:

``` Typescript
import * as crypto from 'crypto';


// example signature from outline webhook:  't=1706609916240,s=45b23d418b5d5c06af7721b526170d23d8699d3de4fc499a21dc33e3576b8236'


// Assuming 'request' is an object with headers and body properties
const outlineSignatureTimestamped = request.headers['outline-signature'];
const timestamp = outlineSignatureTimestamped.split(',')[0].split('=')[1];
const signature = outlineSignatureTimestamped.split(',')[1].split('=')[1];
const body = request.body;

const SECRET = 'YOUR_OUTLINE_SIGNATURE';
const DATA = `${timestamp}.${body.toString()}`; // notice the pattern [timestamp].[StringifiedBody]
const h = crypto.createHmac('sha256', SECRET).update(DATA);
const calculatedSignature = h.digest('hex');

if (calculatedSignature === signature) {
  console.log('Signatures match.😝');
} else {
  console.log('Signatures do not match.😖');
}

**Reply by parsa-asgari** [reply]
The actual point is, to clarify the way the user should verify the request from the outline webhook.
Similar to this: [Slack's Request Verification Docs](https://api.slack.com/authentication/verifying-requests-from-slack)

**Reply by tommoor** [reply]
Thank you for your effort – [the documentation](https://docs.getoutline.com/s/guide/doc/webhooks-gB7HYhS6yq) is now updated based on the example ^
